Cleanup range printing

bhalevy opened the pull request:
This mini-series cleans up printing of ranges in utils/to_string.hh

It generalizes the helper function to work on a std::ranges::range,
with some exceptions, and adds a helper for boost::transformed_range.

It also changes the internal interface by moving `join` the the utils namespace
and use std::string rather than seastar::sstring.

Additional unit tests were added to test/boost/json_test

Fixes #13146

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4923/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4926/

bhalevy commented:
@tchaikov can we close this PR with https://github.com/scylladb/scylladb/pull/13163?

tchaikov commented:
> @tchaikov can we close this PR with #13163?

i will take a closer look tomorrow. sorry for the latency.

tchaikov commented:
@bhalevy i think #13163 does not address all the issues addressed by this one. at least, the commit of "[utils: to_string: generalize range helpers](https://github.com/scylladb/scylladb/pull/13159/commits/47317d61071eb8cc880e369c3bacfb336845004c)" is still relevant before we are in an ideal world. may i rebase your changes onto master HEAD send you a pull request, so we can resurrect it?

bhalevy commented:
> @bhalevy i think #13163 does not address all the issues addressed by this one. at least, the commit of "[utils: to_string: generalize range helpers](https://github.com/scylladb/scylladb/pull/13159/commits/47317d61071eb8cc880e369c3bacfb336845004c)" is still relevant before we are in an ideal world. may i rebase your changes onto master HEAD send you a pull request, so we can resurrect it?

@tchaikov sounds good, thanks!

bhalevy commented:
In v3 ([974f3ea](https://github.com/scylladb/scylladb/commit/974f3ea894adff96d44e0eaac956eabea5a02d30)):
- rebased
- utils: to_string: generalize range helpers
  - simplified, based on fmt::join
- More code removed, and replaced by fmt functionality:
  - utils: to_string: get rid of to_string(const Range&)
  - utils: to_string: get rid of to_string(std::initializer_list)
  - utils: to_string: get rid of utils::join

bhalevy commented:
@tchaikov please re-review

nyh commented on test/boost/json_test.cc:
```diff
@@ -54,3 +68,163 @@ BOOST_AUTO_TEST_CASE(test_parsing_map_from_null) {
+    }
+
+    BOOST_REQUIRE(sv.empty());
+}
```
Philosophical nitpick (feel free to override):
When you write such long utility functions in a test, it becomes a very real possibility that the test function itself has a bug. I admit the possibility is slim (to fail to catch a bug you would need to create a bug in the original code and exactly the same bug in the same code), but I think it would have been easier to just replace:
```c++
    auto vector = std::vector<int>({1, 2, 3});
    test_format_range("vector", vector, std::vector<std::string>({"1", "2", "3"}));
```
By
```
    auto vector = std::vector<int>({1, 2, 3});
    test_format_range("vector", vector, "1, 2, 3"));
```
With two benefits: the utility function is much shorter, and the reader of the test sees immediately what the printed format actually looks like - are there parentheses or not? Which parentheses?

nyh replied:
Oh, what I said here only applies to ordered collections. In a set theoretically you can get any order (although in practice the order would be the same every time). So maybe you need a separate test function for unordered collection. Oh, well... So maybe your approach is as good as it gets.

nyh commented on utils/to_string.hh:
```diff
@@ -9,16 +9,14 @@
+
+#include "bytes.hh"
+
+#include <boost/test/utils/basic_cstring/basic_cstring_fwd.hpp>
```
I don't think we can - or should - use boost/test includes anywhere outside tests. Please don't do it.

nyh commented on utils/to_string.hh:
```diff
@@ -63,25 +61,10 @@ std::ostream& operator<<(std::ostream& os, const print_with_comma<NeedsComma, Pr
 sstring
-to_string(const std::unordered_set<Printable>& items) {
-    return "{" + utils::join(", ", items) + "}";
+to_string(const Range& items) {
```
All the previous functions had "static inline" on them, here you're missing it.
If this is deliberate, please explain why. If not, please add it.

nyh replied:
Also, don't you need all the exclusions below here as well? What if someone does to_string() on something which is already a string, are we ok that we get a list of bytes?

avikivity replied:
`static inline` in a header is bad practice. It leads to duplication of the function, in case the compiler decides not to inline it.

nyh replied:
@avikivity you're probably right, but since this patch changes the practice already in to_string.hh it should be mentioned in the commit message (and please consider if there are any more of them left in to_string.hh).

bhalevy replied:
> All the previous functions had "static inline" on them, here you're missing it. If this is deliberate, please explain why. If not, please add it.

Templates in the header file don't need the `static inline`, which complicates things as Avi said above.

bhalevy replied:
> Also, don't you need all the exclusions below here as well? What if someone does to_string() on something which is already a string, are we ok that we get a list of bytes?

That's why I needed to exclude:
```
+       !std::convertible_to<Range, std::string>
+    && !std::convertible_to<Range, std::string_view>
+    && !std::convertible_to<Range, bytes>
+    && !std::convertible_to<Range, bytes_view>
+    && !std::same_as<Range, boost::unit_test::basic_cstring<const char>>
```
(and `#include <boost/test/utils/basic_cstring/basic_cstring_fwd.hpp>`)).

Otherwise, strings are printed as a range of characters (e.g. "foo" will be printed as `{ 'f', 'o', 'o' }`)
If you insist about not including test header files, I'll try to do that in the test library or find some other solution.

regrading to_string(string), I dropped the `std::to_string` definitions in utils/to_string.hh altogether in the following patches, in favor of fmt::to_string, so they shouldn't matter.

bhalevy replied:
For example, without this, `BOOST_REQUIRE(!sv.empty());` prints the following when running with --log all:
```
test/boost/string_format_test.cc(66): info: check {!, s, v, ., e, m, p, t, y, (, )} has passed
```

nyh commented on utils/to_string.hh:
```diff
@@ -107,45 +90,23 @@ std::ostream& operator<<(std::ostream& os, const std::tuple<T...>& p) {
+template <std::ranges::range Range>
+requires (
+       !std::convertible_to<Range, std::string>
+    && !std::convertible_to<Range, std::string_view>
```
Is sstring "convertible_to" string_view which is why you didn't need to explicitly list it?
If we don't need to explicitly list sstring, maybe we also don't need to explicitly list std::string?

nyh commented on utils/to_string.hh:
```diff
@@ -107,45 +90,23 @@ std::ostream& operator<<(std::ostream& os, const std::tuple<T...>& p) {
+    && !std::convertible_to<Range, std::string_view>
+    && !std::convertible_to<Range, bytes>
+    && !std::convertible_to<Range, bytes_view>
+    && !std::same_as<Range, boost::unit_test::basic_cstring<const char>>
```
Again, I don't like the boost-unit-test specific code here. Why do we need it? Can't we figure out a different way to do this without relying on the specific boost type (e.g., maybe another convertible_to trick?)

tchaikov replied:
`boost::unit_test::basic_cstring` (a.k.a `cstring`) is used in boost::test for the test names and other places. it has its own `operator<<`. to avoid printing it as a range, and to avoid referencing it in `to_string.h`, probably we can have a trait class specialized for it and other range-alike types, as an alternative to type constraints? the trait class for `boost::unit_test::basic_cstring` can be hosted by another header file under `test/` ? but i am wondering what is the use case that we want to format the `cstring` with `seastar::format()`, maybe another option is that we could expose the `cstring` using `seastar::testing::seastar_test` via a `sstring` or a `std::string` ?

bhalevy replied:
We don't format boost_test cstrings using seastar::format.
I think that the compiler picks up our operator<< since they conform to the `std::ranges::range` concept and it calls it implicitly.  Hence this weird exclusion.

nyh commented on test/lib/cql_assertions.cc:
```diff
@@ -38,7 +38,7 @@ rows_assertions::is_empty() {
     if (row_count != 0) {
         auto&& first_row = *rs.rows().begin();
-        fail(format("Expected no rows, but got {:d}. First row: {}", row_count, to_string(first_row)));
+        fail(format("Expected no rows, but got {:d}. First row: {}", row_count, fmt::to_string(first_row)));
```
I don't understand this patch. After you supposedly fixed the to_string() function for ranges in the previous patch, now you don't use it and use fmt::to_string() instead. Why?

nyh replied:
Most importantly, what happens if you DONT do this patch? Does to_string(first_row) do something wrong and fmt::to_string(first_row) fix it? If the answer is yes, we have a big problem because developers are very likely to type to_string() instead of fmt::to_string(), and if it compiles - they won't realize they made a misttake.
If the answer is no, then why is this patch needed?

bhalevy replied:
> I don't understand this patch. After you supposedly fixed the to_string() function for ranges in the previous patch, now you don't use it and use fmt::to_string() instead. Why?

This patch removes the to_string definition from utils/to_string.hh (which makes its name archaic to say the least :)).
I didn't want to do this in a single patch, therefore the separation to patches.
I'll add a FIXME comment in the first patch about this to make the intention clearer.

> Most importantly, what happens if you DONT do this patch? Does to_string(first_row) do something wrong and fmt::to_string(first_row) fix it? If the answer is yes, we have a big problem because developers are very likely to type to_string() instead of fmt::to_string(), and if it compiles - they won't realize they made a misttake. If the answer is no, then why is this patch needed?

We want to implement as little as possible and have a concise implementation.
I'd like the fmt library to do the heavy lifting in this case and use its magic to get to the right formatter.

nyh commented on test/boost/json_test.cc:
```diff
@@ -204,6 +205,9 @@ BOOST_AUTO_TEST_CASE(test_vector_format) {
     test_format_range("initializer_list", std::initializer_list<std::string>{"1", "2", "3"}, std::vector<std::string>({"1", "2", "3"}));
 
+    auto small_vector = utils::small_vector<int, 1>({1, 2, 3});
+    test_format_range("small_vector", small_vector, std::vector<std::string>({"1", "2", "3"}));
```
I don't know why you added this test in this specific patch, but ok :-)

nyh commented on tracing/trace_state.cc:
```diff
@@ -142,7 +142,7 @@ void trace_state::build_parameters_map() {
 
     if (vals.batchlog_endpoints) {
-        params_map.emplace("batch_endpoints", join(sstring(","),  *vals.batchlog_endpoints | boost::adaptors::transformed([](gms::inet_address ep) {return seastar::format("/{}", ep);})));
+        params_map.emplace("batch_endpoints", fmt::to_string(*vals.batchlog_endpoints | boost::adaptors::transformed([](gms::inet_address ep) {return seastar::format("/{}", ep);})));
```
If I understand correctly, this changes the format of this output, adding braces. Is this ok?
I'm not at all a fan of using fmt::to_string with some unspecified formatting for outputting strings which are NOT log messages (as I think is the case here) and where it's not clear that changing the format one day wouldn't break anything.

Can you please make sure that whatever this build_parameters_map() is doing, we have a test for it, so if the format was broken we would know it?

bhalevy replied:
I don't think it matters, but to be on the safe side, I'm just converting this to use fmt::join in v5 so there would be no change in the string format.

nyh reviewed (changes requested):
Looks mostly good, but I left a few comments and requests and what bothered me the most was using Boost Test from within non-test code which I think is not a good idea.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/879/

tchaikov commented on test/boost/json_test.cc:
```diff
@@ -54,3 +68,163 @@ BOOST_AUTO_TEST_CASE(test_parsing_map_from_null) {
+    char open = sv[0];
+    char close = sv[sv.size() - 1];
+    auto it = paren_map.find(open);
+    if (it == paren_map.end()) {
+        BOOST_FAIL(format("Unexpected delimiters: '{}' '{}'", open, close));
+    }
```
nit,
```c++
BOOST_REQUIRE_MESSAGE(
    it != paren_map.end(), 
    format("Unexpected delimiters: '{}' '{}'", open, close));
```

bhalevy replied:
sorry, I think it's less intuitive.
I don't see the value in this macro.

tchaikov commented on test/boost/json_test.cc:
```diff
@@ -54,3 +68,163 @@ BOOST_AUTO_TEST_CASE(test_parsing_map_from_null) {
+} // namespace
+
+BOOST_AUTO_TEST_CASE(test_vector_format) {
+    auto vector = std::vector<int>({1, 2, 3});
```
nit, could put something like:

```c++
auto ints = {1, 2, 3};
auto strings = std::vector<std::string_view>{"1", "2", "3"};
test_format_range("vector", std::vector(ints), strings);
```

bhalevy replied:
hmm, looks nice.
I'll try to apply this approach.

tchaikov commented on test/boost/json_test.cc:
```diff
@@ -54,3 +68,163 @@ BOOST_AUTO_TEST_CASE(test_parsing_map_from_null) {
+
+template <std::ranges::range Range>
+void test_format_range(const char* desc, Range x, std::unordered_set<std::string> expected_strings, std::string expected_delim = ",", bool expect_parenthesis = true) {
+    auto str = seastar::format("{}", x);
```
it seems we are testing `seastar::format()` here, which is a very thing wrapper around `fmt::format_to()` . so probably we could have a short note before these tests to explain how/why we need them in scylladb. without more context, a reader might wonder why we would have them here. in other words, it's like testing abseil in scylladb. for instance, how are they related to json related processing?

tchaikov commented on test/boost/json_test.cc:
```diff
@@ -54,3 +68,163 @@ BOOST_AUTO_TEST_CASE(test_parsing_map_from_null) {
+    auto unordered_map = std::unordered_map<int, std::string>({{1, "one"}, {2, "two"}, {3, "three"}});
+    // seastar has a specialized print function for unordered_map
+    // See https://github.com/scylladb/seastar/issues/1544
+    test_format_range("unordered_map", unordered_map, std::unordered_set<std::string>({"{1 -> one}", "{2 -> two}", "{3 -> three}"}));
```
IMHO, the notation is kind of arbitrary, and it is not the same as how fmtlib formats `std::unordered_map`. do we rely on this formatting, like we do in dtest?

bhalevy replied:
I don't think we do, but it's kind of the convention now in scylla.
We use it to print e.g. option dictionaries (`std::unordered_map<sstring, sstring>`)

tchaikov replied:
thank you for confirming this. I was asking because I was actually thinking about dropping it, and replacing it with fmtlib. Will keep your answer in mind.

tchaikov commented on utils/to_string.hh:
```diff
@@ -107,45 +90,23 @@ std::ostream& operator<<(std::ostream& os, const std::tuple<T...>& p) {
-std::ostream& operator<<(std::ostream& os, const std::list<T>& items) {
-    os << "[" << utils::join(", ", items) << "]";
+template <typename... Args>
+std::ostream& operator<<(std::ostream& os, const boost::transformed_range<Args...>& items) {
```
didn't know this. this is neat!

bhalevy replied:
Yeah. it is really convenient.

avikivity commented on utils/to_string.hh:
```diff
@@ -63,25 +61,10 @@ std::ostream& operator<<(std::ostream& os, const print_with_comma<NeedsComma, Pr
-    return "{" + utils::join(", ", items) + "}";
+to_string(const Range& items) {
+    return fmt::format("{{{}}}", fmt::join(items, ", "));
 }
```
Should we even have to_string(range)? Isn't it better for each call site to decide on its own how to print it?

avikivity replied:
Note that within a formatting snippet, fmt::join() is more efficient since it never allocates an intermediate string, but using to_string() does.

bhalevy replied:
Agreed.  those to_string definitions are gone in the next couple patches.
I'll add a FIXME in the next version so it would be more evident.
I just didn't want to do both in a single patch, which is complex enough as it is.

bhalevy commented on test/boost/expr_test.cc:
```diff
@@ -1,11 +1,20 @@
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#define BOOST_TEST_MODULE core
```
sorry, this patch was left over from the previous version and is no longer required.
will remove in next version.

bhalevy commented:
In v4 ([52018f0](https://github.com/scylladb/scylladb/commit/52018f0d0b14c11dc6aef51e7801488e90c6d870)):
- test: add string_format_test
  - added unit test upfront
- utils: to_string: generalize range helpers
  - opt-in instead of opt-out
  - template specialization on `std::ranges::range` alone was too inclusive and caused regressions with boost strings
  - add common `utils::format_range` helper
  - added/converted printers for `utils/{small,chunked}_vector` using the common helper.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/936/

tchaikov commented on test/boost/string_format_test.cc:
```diff
@@ -0,0 +1,265 @@
+
+/*
+ * Copyright (C) 2015-present ScyllaDB
```
s/2015/2023/

bhalevy replied:
fixed in next version

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/936/

#13211

bhalevy commented:
[613fe9f](https://github.com/scylladb/scylladb/commit/613fe9f4c86a7665b3ee741a526e068fca800800): fixed copyright year

bhalevy commented:
rebased

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/959/

bhalevy commented:
@scylladb/scylla-maint I'm not sure why this PR still shows up with "Changes requested".
AFAIU all review comments were address.
Please consider merging.

denesb reviewed (approved):
LGTM, but there is an unanswered comment from @nyh on the last commit.

bhalevy commented:
In v5 ([959a740](https://github.com/scylladb/scylladb/commit/959a740dac56763a53e7a2fd574f576eef4574c9)):
- Addressed https://github.com/scylladb/scylladb/pull/13159/commits/deddd70b921bd4f0d48ee524546bb631e7e06b22#r1176252843
  - kept the existing string format in `trace_state::build_parameters_map` for `batch_endpoints`, just using `fmt::format` + `fmt::join`.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1012/
